### ISSUE TITLE
import binding not linked in `define` function

### ISSUE THREAD

**Issue by magic-akari** [webpack/webpack#17063]
# Bug report

Keywords:
- AMD
- define
- import

**What is the current behavior?**

import binding in define function will cause referenceError.

**If the current behavior is a bug, please provide the steps to reproduce.**

Playground: https://stackblitz.com/edit/node-y2sktu?file=dist/main.js&view=editor

Input:

```JavaScript
// lib.js
export default 42;

// index.js
import foo from './lib.js';
const bar = require('./lib.js');

define(function (require) {
  console.log(foo); // not work
  console.log(bar); // ok
});

noop(function () {
  console.log(foo); // ok
  console.log(bar); // ok
});

function noop() {}

```

output:

```JavaScript
// omitted

/* harmony import */ var _lib_js__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./lib.js */ "./src/lib.js");

const bar = __webpack_require__(/*! ./lib.js */ "./src/lib.js");

define(function (require) {
  console.log(foo); // not work
  console.log(bar); // ok
});

noop(function () {
  console.log(_lib_js__WEBPACK_IMPORTED_MODULE_0__["default"]); // ok
  console.log(bar); // ok
});

function noop() {}

// omitted
```

**What is the expected behavior?**

I am not sure if it's a bug.
But I expect that `foo` should be transpiled into `_lib_js__WEBPACK_IMPORTED_MODULE_0__["default"]`.

**Other relevant information:**
webpack version:  5.38.1
Node.js version: 16.20.0
Operating System: Darwin arm64
Additional tools:

related issues:
- #14307 
- swc-project/swc/issues/7334

**Comment by alexander-akait** [comment]
yeah, AMD again, not sure why developers still use it in 2023, feel free to send a fix

**Comment by magic-akari** [comment]
I am willing to fix it.

**Comment by alexander-akait** [comment]
bump

**Comment by magic-akari** [comment]
Despite several attempts, I did not get the expected result. 
If someone is ready to work on it, go ahead.

**Comment by hai-x** [comment]
I found that for `define` callExpression. Its args will skip parsing in harmony module that causes import binding can't be linked.

https://github.com/webpack/webpack/blob/964c0315df0ee86a2b4edfdf621afa19db140d4f/lib/dependencies/HarmonyDetectionParserPlugin.js#L104

**Comment by hai-x** [comment]
> I found that for `define` callExpression. Its args such as `(function (require) {console.log(foo); // not work})` will skip parsing in harmony module that causes import binding can't be linked.
> 
> https://github.com/webpack/webpack/blob/964c0315df0ee86a2b4edfdf621afa19db140d4f/lib/dependencies/HarmonyDetectionParserPlugin.js#L104
> 
> webpack@4 also have this logic.
> 
> https://github.com/webpack/webpack/blob/d4603c6d7ef4acfbe956ef49d8476c7368526989/lib/dependencies/HarmonyDetectionParserPlugin.js#L77
> 
> The difference is webpack4 will process VarDeclaration defined in top-level. It may be caused by `innerGraph`.

```js
// lib.js
export default 42;

// index.js
import foo from './lib.js';

let d = function (require) {
  console.log(foo);
}

define(d);
```
and  `optimization.innerGraph: false`. 

It works.

**Comment by hai-x** [comment]
After look into this issue in depth. 

I found the issue is caused by `innerGraph`. 

`foo` dep.usedExport will be set to false which makes 2 module can't be connected. So we can temporarily workaround it by setting `optimization.innerGraph: false`

Meanwhile, there is another problem. We need to declare the arguments of `define` function call at top-level. Otherwise, it will skip parsing of callExpression.arguments by `HarmonyDetectionParserPlugin`. https://github.com/webpack/webpack/blob/964c0315df0ee86a2b4edfdf621afa19db140d4f/lib/dependencies/HarmonyDetectionParserPlugin.js#L104

Following can work.

```js
// lib.js
export default 42;

// index.js
import foo from './lib.js';
const bar = require('./lib.js');

const arg = function (require) {
  console.log(foo); // not work
  console.log(bar); // ok
}

define(arg);

noop(function () {
  console.log(foo); // ok
  console.log(bar); // ok
});

function noop() {}

// webpack.config.js
module.exports = {
	// ...
	optimization:{
		// ...
		innerGraph: false
	}
}
```

**Comment by aryanraj45** [comment]
Implemented a fix based on the analysis by @hai-x added a stage: -1 tap on hooks.call.for("define") to walk expression.arguments before skipInHarmony (stage 0) bails. This also fixes the innerGraph case where foo.usedExport was incorrectly set to false.
